csrc/quantization/quant_llm/quant_llm_linear.cu

```diff
@@ -0,0 +1,301 @@
+W * trans(In). Note that we do not perform "transpose" during runtime, we
+instead interpret the In/Out as column-major matrices when calling our CUDA
+kernel. [Inputs] _in_feats:  tensor of shape [B, IC];                  // half
+  _weights:   int tensor of shape [OC, IC // 8 * x];    // x UINT8 words
+contains 8 FPx weights. _scales:    tensor of shape [OC];                     //
+half splitK:     splitting the MatMul problem along K dimension for higher GPU
+utilization, default 1. [Outputs] _out_feats: tensor of shape [B, OC]; // half
+*/
+torch::Tensor fp_eXmY_linear_forward_cuda(int64_t EXPONENT, int64_t MANTISSA,
+                                          torch::Tensor _in_feats,
+                                          torch::Tensor _weights,
+                                          torch::Tensor _scales,
+                                          int64_t splitK = 1) {
+  const int64_t NBITS = 1 + EXPONENT + MANTISSA;
+  int num_in_feats = _in_feats.size(0);
+  int num_in_channels = _in_feats.size(1);
+  int num_out_channels = _weights.size(0);
+  TORCH_CHECK(num_in_channels % 64 == 0,
+              "Expected in_features to be a multiple of 64, but received ",
+              num_in_channels);
+  TORCH_CHECK((num_in_channels / 8 * NBITS) ==
+              _weights.size(1));  // Making sure the K dimension is matched.
+  //
+  int M = num_out_channels;
+  int K = num_in_channels;
+  int N = num_in_feats;
+  // Input Tensors
+  auto weight = reinterpret_cast<const uint4*>(
+      _weights.data_ptr<uint8_t>());  // weights is [OC, IC] but in FP6.
+  auto in_feats = reinterpret_cast<const half*>(_in_feats.data_ptr<at::Half>());
+  auto scales = reinterpret_cast<const half*>(_scales.data_ptr<at::Half>());
+  // Output Tensors
+  auto options = torch::TensorOptions()
+                     .dtype(_in_feats.dtype())
+                     .device(_in_feats.device());
+  at::Tensor _out_feats =
+      torch::empty({num_in_feats, num_out_channels}, options);
+  auto out_feats = reinterpret_cast<half*>(_out_feats.data_ptr<at::Half>());
+
+  options =
+      torch::TensorOptions().dtype(torch::kFloat32).device(_in_feats.device());
+  at::Tensor _workspace =
+      torch::empty({splitK, num_in_feats, num_out_channels}, options);
+  auto Reduction_Workspace = reinterpret_cast<float*>(
+      _workspace.data_ptr<float>());  // Reduction_Workspace_Size = Split_K *
+                                      // M_Global * N_Global * sizeof(fp32)
+
+  // NOTE(alpin): use at::cuda::getCurrentCUDAStream() instead of default
+  // stream (0) this fixes problem with CUDA graphs when used with
+  // torch.compile()
+  auto stream = at::cuda::getCurrentCUDAStream();
+
+  /*
+   The heuristic is weight_bit - exponent_bit - 1 = mantissa_bit
+
+   NOTE(alpin): Using switch-statement here probably doesn't matter,
+   the compiler will likely optimize it to a jump table.
+   */
+
+  // FP4
+  else if (EXPONENT == 1 && MANTISSA == 2)
```

Comment:
I got a build error here. It looks like this should be `if` instead of `else if` ?

```suggestion
 if (EXPONENT == 1 && MANTISSA == 2)
```

Reply:
Yes, thanks for catching that. Didn't notice when I removed FP2/FP3 earlier.